### PR TITLE
Add informer for Service definitions

### LINKED ISSUE #61: Services are not re-created when deleted manually
## Expected Behaviour

If someone deletes a resource which belongs to the operator, then the operator should re-create it - whether that be the deployment or the service entry.

## Current Behaviour

It doesn't - it won't re-create the ClusterIP service.

## Steps to Reproduce (for bugs)

```
alex@nuc7:~/go/src/github.com/openfaas/certifier$ faas-cli store deploy figlet -g 127.0.0.1:31112

Deployed. 202 Accepted.
URL: http://127.0.0.1:31112/function/figlet

alex@nuc7:~/go/src/github.com/openfaas/certifier$ kubectl get svc,deploy -n openfaas-fn
NAME      TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
figlet    ClusterIP   10.96.217.119   <none>        8080/TCP   6s

NAME      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
figlet    1         1         1            0           6s
alex@nuc7:~/go/src/github.com/openfaas/certifier$ kubectl delete svc/figlet -n openfaas-fn
service "figlet" deleted
alex@nuc7:~/go/src/github.com/openfaas/certifier$ kubectl get svc,deploy -n openfaas-fn
NAME      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
figlet    1         1         1            1           17s
alex@nuc7:~/go/src/github.com/openfaas/certifier$ kubectl get svc,deploy -n openfaas-fn
NAME      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
figlet    1         1         1            1           19s
alex@nuc7:~/go/src/github.com/openfaas/certifier$ kubectl get svc,deploy -n openfaas-fn
NAME      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
figlet    1         1         1            1           22s
alex@nuc7:~/go/src/github.com/openfaas/certifier$ kubectl get svc,deploy -n openfaas-fn
NAME      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
figlet    1         1         1            1           23s
alex@nuc7:~/go/src/github.com/openfaas/certifier$ 
```

These are the logs that we get:

```
I1202 20:42:50.747457       1 apply.go:65] Function figlet created
I1202 20:42:50.752514       1 controller.go:300] Creating ClusterIP service for 'figlet'
I1202 20:42:50.760943       1 controller.go:315] Creating deployment for 'figlet'
E1202 20:42:50.764466       1 controller.go:257] error syncing 'openfaas-fn/figlet': deployment.apps "figlet" not found
```

The error `'openfaas-fn/figlet': deployment.apps "figlet" not found` seems to be a false error since the CRD entry still exists and the deployment.

If I delete the `Deployment` then the `Deployment` and the `Service` are both re-created.

## Context

This seems like a fringe-case, but one we'd need to cover for automation.

## Your Environment

* Docker version `docker version` (e.g. Docker 17.0.05 ):

```
 Version:	18.04.0-ce
```

* What version and distriubtion of Kubernetes are you using? `kubectl version`

```
alex@nuc7:~/go/src/github.com/openfaas/certifier$ kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:55:54Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.8", GitCommit:"7eab6a49736cc7b01869a15f9f05dc5b49efb9fc", GitTreeState:"clean", BuildDate:"2018-09-14T15:54:20Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

* Operating System and version (e.g. Linux, Windows, MacOS):

Ubuntu Linux 16.04

* What network driver are you using and what CIDR? i.e. Weave net / Flannel

Weave

### PR DESCRIPTION
**Pull request by alexellis**
Add informer for Service definitions

Signed-off-by: Alex Ellis (VMware) <alexellis2@gmail.com>

## Description

- adds informer for Service definitions to fix #61 and #60 -
since it seemed that deleting Service definitions was not tracked
or synchronized and the code was depending on the Deployment
from being deleted or updated.

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

Fixes #61 and #60

## How Has This Been Tested?

- [ ] I have run the [certifier](https://github.com/openfaas/certifier)

Tested on K8s 1.10 on single node Linux cluster by deleting
Service and seeing it re-created in the logs. Happy path of
deploying from store also works unaffected.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Impact to existing users

Improved durability when resources owned by controller are deleted manually or accidentally by automation.

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -154,6 +157,21 @@ func NewController(
 		DeleteFunc: controller.handleObject,
 	})
 
+	serviceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.handleObject,
+		UpdateFunc: func(old, new interface{}) {
+			newDepl := new.(*corev1.Service)
```
Comment (LucasRoesler) [inline pkg/controller/controller.go]: looks like some copy-pasta here in the variable name.  Should be `newSvc` and `oldSvc`.  Same for the comment, it also references Deployments

Comment (alexellis) [reply pkg/controller/controller.go]: Fair point.

**Comment by stefanprodan** [comment]
There is no need for handling updates anyway, since the code only deals with creating a service if it doesn’t exists

**Comment by alexellis** [comment]
@stefanprodan do you mean we can drop `AddFunc` and `UpdateFunc` just leaving the handler for `DeleteFunc`?

**Comment by alexellis** [comment]
Or should we leave the subscription and in the future remediate any changes made to Services that are under our control (by ownership reference)?

**Comment by LucasRoesler** [comment]
Is there a concrete change being suggested, should `UpdateFunc` be `nil` or implement 3-way merge before we merge this PR? Seems like setting it to `nil` would be safest for now.

**Comment by alexellis** [comment]
Given 2x approvals I'll merge, happy to adapt in new PR if needed.

**Pull request merged**